### PR TITLE
Add 3D Map View toolbars and animation

### LINKED ISSUE #2140: [FEATURE] QGIS 3D (pull request #5203)
Original commit: https://github.com/qgis/QGIS/commit/68bb68d17682937bb8a9d1b26968303dc0de87fb by web-flow

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #2979: [FEATURE] Identification map tool for 3D views
Original commit: https://github.com/qgis/QGIS/commit/407a5cb51eaffc64da4f598432b33f15c0fdca28 by wonder-sk

Currently identification only works on terrain and does not highlight in 3D

### LINKED ISSUE #2997: [FEATURE] 3D identify tool working on 3D entities
Original commit: https://github.com/qgis/QGIS/commit/2085dfa6aefc39800f453f2ab4a1242568701817 by wonder-sk

Until now the tool only considered terrain. This commit adds support
for 3D renderers created from vector layers, so it is possible to
correctly identify polygons and linestrings in 3D.

### LINKED ISSUE #3476: [3d] export all frames from QGIS 3d animations as images (#9244)
Original commit: https://github.com/qgis/QGIS/commit/209e9144f87654ac6dc11dadfc539d387d571ece by web-flow

[feature] [3d] export all frames from QGIS 3d animations as images #21300

### PR DESCRIPTION
**Pull request by DelazJ**
Fixes #2979, fixes #2997, fixes #3476, refs #3970 and let's close #2140

### CODE REVIEW

File: docs/user_manual/introduction/qgis_gui.rst [hunk 1]
```diff
@@ -2376,21 +2376,38 @@ In the dialog that opens:
...
+the view into 3D.
+
+.. _figure_3dmapview:
+
+.. figure:: img/3dmapview.png
+   :align: center
+
+   The 3D Map View dialog
+
+Following tools are provided at the top of the 3D map view panel:
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2394]:
```suggestion
The following tools are provided at the top of the 3D map view panel:
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 2]
```diff
@@ -2376,21 +2376,38 @@ In the dialog that opens:
...
+
+* |pan| :sup:`Camera control`: moves the view, keeping the same angle
+  and direction of the camera
+* |zoomFullExtent| :sup:`Zoom Full`: resizes the view to the whole
+  layers' extent
+* |3dNavigation| :sup:`Toggle on-screen notification`: shows/hides the
+  navigation widget for a better control of the map view
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2401]:
```suggestion
  navigation widget
```
As it was written, it does not say whether it is with or without this navigation widget that the control of the map view is better.

File: docs/user_manual/introduction/qgis_gui.rst [hunk 3]
```diff
@@ -2376,21 +2376,38 @@ In the dialog that opens:
...
+* |identify| :sup:`Identify`: returns attributes of the clicked feature(s)
+  -- More details at :ref:`identify`
+* |measure| :sup:`Measurement line`: measures horizontal distance between points
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2404]:
```suggestion
* |measure| :sup:`Measurement line`: measures the horizontal distance between points
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 4]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+---------------------
+
+An animation is based on a set of keyframes - camera positions at particular times.
+To create an animation:
+
+#. Toggle on the |play| :sup:`Animations` tool, displaying the animation player
+   widget
+#. Click the |signPlus| :sup:`Add keyframe` button and enter a :guilabel:`Keyframe
+   time` in seconds. The :guilabel:`Keyframe` combo box now displays the time set.
+#. Using the navigation tools, move the camera to a position to associate to
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2474]:
```suggestion
#. Using the navigation tools, move the camera to the position to associate with
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 5]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+   the current keyframe time.
+#. Repeat previous steps to add as many keyframes time and position as necessary.
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2476]:
```suggestion
#. Repeat the previous steps to add as many keyframes (with time and position) as necessary.
```
Or skip ` (with time and position)` altogether.

File: docs/user_manual/introduction/qgis_gui.rst [hunk 6]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+#. Click |play| button to preview the animation. QGIS will generate a scene using
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2477]:
```suggestion
#. Click the |play| button to preview the animation. QGIS will generate scenes using
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 7]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+   the camera positions/rotations at set times and interpolating them in between
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2478]:
```suggestion
   the camera positions/rotations at set times, and interpolating them in between
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 8]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+   these keyframes. Various :guilabel:`Interpolation` modes for animations are
+   available (eg, linear, inQuad, outQuad, inCirc... -- more details at
+   https://doc.qt.io/qt-5/qeasingcurve.html#EasingFunction-typedef).
+
+   The animation can also be previewed by moving the time slider.
+   Keeping the |draw| :sup:`Repeat` button pressed will repeatedly run the
+   animation while clicking |play| stops a running animation.
+
+It's possible to browse the different views of the camera, using the
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2487]:
```suggestion
It is possible to browse the different views of the camera, using the
```

File: docs/user_manual/introduction/qgis_gui.rst [hunk 9]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+:guilabel:`Keyframe` list. Whenever a time is active, changing the map view
+will automatically update the associated position. You can also |symbologyEdit|
+:sup:`Edit keyframe` time or |signMinus| :sup:`Remove keyframe`.
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2490]:
```suggestion
:sup:`Edit keyframe` or |signMinus| :sup:`Remove keyframe`.
```
Is `time` necessary here?

File: docs/user_manual/introduction/qgis_gui.rst [hunk 10]
```diff
@@ -2442,6 +2459,43 @@ Navigation options for exploring the map in 3D:
...
+
+Click |fileSave| :sup:`Export animation frames` to generate a series of image
```
Comment (havatv) [inline docs/user_manual/introduction/qgis_gui.rst:2492]:
```suggestion
Click |fileSave| :sup:`Export animation frames` to generate a series of images
```

File: docs/user_manual/introduction/qgis_gui.rst (hunk 2 above)
Comment (DelazJ) [reply docs/user_manual/introduction/qgis_gui.rst:2401]: Would
```suggestion
  navigation widget that can be used for a better control of the map view
```
(or replace "a better" with "an easy" or "a convenient") make it clearer? I think the navigation widget is meant to ease the navigation, compared to the previous "guess yourself how to proceed"

File: docs/user_manual/introduction/qgis_gui.rst (hunk 9 above)
Comment (DelazJ) [reply docs/user_manual/introduction/qgis_gui.rst:2490]: This button allows to change the time only. Position is not concerned.

File: docs/user_manual/introduction/qgis_gui.rst (hunk 2 above)
Comment (havatv) [reply docs/user_manual/introduction/qgis_gui.rst:2401]: I  tried to modify.

File: docs/user_manual/introduction/qgis_gui.rst (hunk 9 above)
Comment (havatv) [reply docs/user_manual/introduction/qgis_gui.rst:2490]: OK. I tried to rephrase.

**Pull request merged**